### PR TITLE
[frontend-api] add display timezone to SettingsQuery

### PR DESCRIPTION
**Pull request by vitek-rostislav**
| Q             | A
| ------------- | ---
|Description, reason for the PR| For proper formatting on the storefront, the API must provide the display timezone, plus all the dates must be stored with time.
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

----
:globe_with_meridians: Live Preview:
  - https://rv-display-timezone.odin.shopsys.cloud
  - https://cz.rv-display-timezone.odin.shopsys.cloud

### CODE REVIEW

File: packages/framework/src/Model/Store/ClosedDay/ClosedDay.php [hunk 1]
```diff
@@ -36,7 +36,7 @@ class ClosedDay
 
     /**
      * @var \DateTimeImmutable
-     * @ORM\Column(type="date_immutable")
+     * @ORM\Column(type="datetime_immutable")
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/ClosedDay/ClosedDay.php]: the date field in form also have the `'view_timezone' => DateTimeHelper::UTC_TIMEZONE,` set.

Maybe it would be nice to search the codebase for usage of the UTC timezone

File: packages/frontend-api/src/Model/Resolver/Settings/DisplayTimezoneQuery.php [hunk 2]
```diff
@@ -22,6 +22,6 @@ public function __construct(
      */
     public function displayTimezoneQuery(): string
     {
-        return $this->displayTimeZoneProvider->getDisplayTimeZone()->getName();
+        return $this->domain->getDateTimeZone()->getName();
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Resolver/Settings/DisplayTimezoneQuery.php]: Shouldn't it be the other way around? I would expect the display timezone provider to take into account the timezone settings in domains

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/frontend-api/src/Model/Resolver/Settings/DisplayTimezoneQuery.php (hunk 2 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Resolver/Settings/DisplayTimezoneQuery.php]: re-worked in 60e23b277

Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Resolver/Settings/DisplayTimezoneQuery.php]: and 733dafcc0 :slightly_smiling_face:

File: packages/framework/src/Model/Store/ClosedDay/ClosedDay.php (hunk 1 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/ClosedDay/ClosedDay.php]: resolved in c45327a25

Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/ClosedDay/ClosedDay.php]: resolved in c45327a25

File: UPGRADE-14.0.md [hunk 3]
```diff
@@ -769,6 +769,21 @@ Follow the instructions in relevant sections, e.g. `shopsys/coding-standards` or
 -   add Category ID to CategoryHierarchyItem ([#2962](https://github.com/shopsys/shopsys/pull/2962))
     -   see #project-base-diff to update your project
 -   add display timezone to FE API SettingsQuery ([#2977](https://github.com/shopsys/shopsys/pull/2977))
+    -   `Shopsys\FrameworkBundle\Component\DateTimeHelper\DateTimeHelper` class was changed:
+        -   `UTC_TIMEZONE` constant was removed
+        -   `TIME_REGEX` constant was removed
+        -   `convertDatetimeStringFromDisplayTimeZoneToUtc()` method was removed
+        -   `convertDateTimeFromUtcToDisplayTimeZone()` method was removed
```
Comment (grossmannmartin) [inline UPGRADE-14.0.md]: everything was added in this version, so I think the upgrade notes are not necessary

**Review by grossmannmartin — APPROVED** [review]

File: UPGRADE-14.0.md (hunk 3 above)
Comment (vitek-rostislav) [reply UPGRADE-14.0.md]: ah, good catch :ok_hand: fixed in f1a124537

File: packages/framework/src/Form/Admin/Store/StoreFormType.php [hunk 4]
```diff
@@ -182,6 +197,7 @@ public function configureOptions(OptionsResolver $resolver): void
             ->setDefaults([
                 'data_class' => StoreData::class,
                 'attr' => ['novalidate' => 'novalidate'],
+                'constraints' => new Constraints\Callback($this->validateOpeningHours(...)),
```
Comment (grossmannmartin) [inline packages/framework/src/Form/Admin/Store/StoreFormType.php]: this callback validator breaks any js validation, or?

File: packages/framework/src/Form/DatePickerType.php [hunk 5]
```diff
@@ -15,9 +15,9 @@ class DatePickerType extends AbstractType
     public const FORMAT_JS = 'dd.mm.yy';
 
     /**
-     * @param \Shopsys\FrameworkBundle\Component\Localization\DisplayTimeZoneProviderInterface|null $displayTimeZoneProvider
+     * @param \Shopsys\FrameworkBundle\Component\Localization\DisplayTimeZoneProvider $displayTimeZoneProvider
```
Comment (grossmannmartin) [inline packages/framework/src/Form/DatePickerType.php]: I think it should be annotated with interface too instead of custom implementation

File: packages/framework/src/Form/DatePickerType.php [hunk 6]
```diff
@@ -30,12 +30,9 @@ public function configureOptions(OptionsResolver $resolver)
             'widget' => 'single_text',
             'format' => static::FORMAT_PHP,
             'html5' => false,
+            'display_timezone' => $this->displayTimeZoneProvider->getDisplayTimeZoneForAdmin(),
```
Comment (grossmannmartin) [inline packages/framework/src/Form/DatePickerType.php]: display_timezone or view_timezone?

File: packages/framework/src/Migrations/Version20240130104139.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+class Version20240130104139 extends AbstractMigration
+{
+    /**
+     * @param \Doctrine\DBAL\Schema\Schema $schema
+     */
+    public function up(Schema $schema): void
+    {
+        $this->sql('ALTER TABLE stores ADD domain_id INT NOT NULL DEFAULT 1');
+        $storeDomainsData = $this->sql('SELECT store_id, domain_id FROM store_domains WHERE is_enabled = TRUE ORDER BY domain_id')->fetchAllAssociative();
```
Comment (grossmannmartin) [inline packages/framework/src/Migrations/Version20240130104139.php:18]: storedomain db table should remain?

File: packages/framework/src/Migrations/Version20240131163656.php [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+            if ($firstOpeningTime !== null && $firstClosingTime !== null) {
+                $this->insertIntoOpeningHours($firstOpeningTime, $firstClosingTime, $storeId, $dayOfWeek);
+            }
+
+            if ($secondOpeningTime !== null && $secondClosingTime !== null) {
+                $this->insertIntoOpeningHours($secondOpeningTime, $secondClosingTime, $storeId, $dayOfWeek);
+            }
+        }
+        $this->sql('DELETE FROM store_opening_hours WHERE opening_time IS NULL AND closing_time IS NULL');
```
Comment (grossmannmartin) [inline packages/framework/src/Migrations/Version20240131163656.php]: this may be null at any point?

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/framework/src/Form/Transformers/OpeningHoursCollectionTransformer.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        return OpeningHoursDataHelper::getOpeningHoursIndexedByDayNumber($value);
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Store\OpeningHours\OpeningHoursData[][] $value
+     * @return \Shopsys\FrameworkBundle\Model\Store\OpeningHours\OpeningHoursData[]
+     */
+    public function reverseTransform($value)
+    {
+        $flattenedOpeningHours = [];
```
Comment (grossmannmartin) [inline packages/framework/src/Form/Transformers/OpeningHoursCollectionTransformer.php]: does it make sense to have this code also in helper?

File: packages/framework/src/Model/Store/StoreDataFactory.php [hunk 10]
```diff
@@ -74,12 +73,14 @@ public function createFromStore(Store $store): StoreData
     }
 
     /**
+     * @param int $domainId
      * @return \Shopsys\FrameworkBundle\Model\Store\StoreData
      */
-    protected function createInstance(): StoreData
+    protected function createInstance(int $domainId): StoreData
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/StoreDataFactory.php]: createInstance should only create the object. Rest of the values set should be in setData or create/createfromstore methods. 
Currently, for example, data are set and are immediately overridden in createFromStore.

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    ) {
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Store\Store $store
+     * @return bool
+     */
+    public function isOpenNow(Store $store): bool
+    {
+        $now = (new DateTimeImmutable())->setTimezone($this->displayTimeZoneProvider->getDisplayTimeZoneByDomainId($store->getDomainId()));
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: nitpick:

may be written as 
```php
$now = new DateTimeImmutable(timezone: $this->displayTimeZoneProvider->getDisplayTimeZoneByDomainId($store->getDomainId()));
```

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        4 => 'thursday',
+        5 => 'friday',
+        6 => 'saturday',
+        7 => 'sunday',
+    ];
+
+    /**
+     * @var \Spatie\OpeningHours\OpeningHours[]
+     */
+    protected array $openingHoursSetting = [];
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php:28]: Should this provider implement the ResetInterface? Or it's ok to keep this setting during whole message processing?

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+     */
+    protected function addOpeningHourRangeToWeekSetting(
+        int $dayOfWeek,
+        ?string $openingTime,
+        ?string $closingTime,
+        array $weekSetting,
+    ): array {
+        if ($openingTime !== null && $closingTime !== null) {
+            $openingHoursRange = $openingTime . '-' . $closingTime;
+            $weekSetting[$this->getEnglishDayNameFromDayNumber($dayOfWeek)][] = $openingHoursRange;
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: Does it make sense to keep the opening and closing time nullable when they are ignored in that case? 
What would change if opening and closing times were non-nullable in entity (and in higher layers)

Then this method will maybe just `formatOpeningHourRange(string $openingTime, string $closingTime)`

and the usage will be 
```
$weekSetting[$this->getEnglishDayNameFromDayNumber($dayOfWeek)][] = formatOpeningHours()
```

I find it more understandable than passing array that is modified and returned

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Store\OpeningHours\OpeningHoursData[] $openingHoursData
+     * @return \Spatie\OpeningHours\OpeningHours
+     */
+    public function getOpeningHoursSettingFromData(array $openingHoursData): SpatieOpeningHours
+    {
+        $weekSetting = [];
+
+        foreach ($openingHoursData as $openingHourData) {
+            $weekSetting = $this->addOpeningHourRangeToWeekSetting($openingHourData->dayOfWeek, $openingHourData->openingTime, $openingHourData->closingTime, $weekSetting);
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: nitpick

I think the arguments may be on separate lines to improve readability

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Store\Store $store
+     * @return array[][]
+     */
+    protected function getExceptions(Store $store): array
+    {
+        $exceptions = [];
+        $closedDays = $this->closedDayFacade->getThisWeekClosedDaysNotExcludedForStore(
+            $this->domain->getId(),
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: Why current domain id? The whole provider seems independent of the application state. 
Couldn't it be `$store->getDomainId()`?

maybe then the repository method may be simplified too, I'm not sure now

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+            $openingHourData = $this->openingHoursDataFactory->create();
+            $openingHourData->dayOfWeek = $this->getDayNumberFromEnglishDayName($dayName);
+
+            return [$openingHourData];
+        }
+
+        $openingHoursData = [];
+        /** @var \Spatie\OpeningHours\TimeRange $openingHour */
+        foreach ($openingHoursForDay->getIterator() as $openingHour) {
+            $openingHourData = $this->openingHoursDataFactory->create();
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: nitpick

Just thinking whether it would make sense to change signature of create method to 
`create(int $dayOfWeek)`

it seems like it has to be always constructed with the dayOfWeek or the object is invalid

File: packages/framework/src/Model/Store/StoreFacade.php [hunk 17]
```diff
@@ -192,4 +178,47 @@ public function getStoresByIds(array $storeIds): array
...
+        $daysOfWeek = range(1, 7);
+        $missingDays = array_diff($daysOfWeek, $daysCovered);
+
+        foreach ($missingDays as $missingDay) {
+            $openingHourData = $this->openingHoursDataFactory->create();
+            $openingHourData->dayOfWeek = $missingDay;
+            $openingHours[] = $this->openingHoursFactory->create($openingHourData);
+        }
+
+        return array_map(
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/StoreFacade.php]: usually, I like array_map, but in this case wouldn't be this more readable?

```
foreach ($openingHours as $openingHour) {
    $openingHour->setStore($store);
}

return $openingHours;
```

maybe even better, the store may be set directly in both loops, or the openingHoursFactory may contain the method `createWithStore(OpeningHourData $data, Store $store)`

File: project-base/app/tests/FrontendApiBundle/Functional/Order/TransportInOrderValidationTest.php [hunk 18]
```diff
@@ -108,28 +102,6 @@ public function testTransportPriceChanged(): void
         $this->assertTrue($response['data']['CreateOrder']['cart']['modifications']['transportModifications']['transportPriceChanged']);
     }
 
-    public function testDisabledPickupPlaceUnavailable(): void
```
Comment (grossmannmartin) [inline project-base/app/tests/FrontendApiBundle/Functional/Order/TransportInOrderValidationTest.php:111]: do we want to test the pickup place from another domain? or it's tested elsewhere?

File: packages/framework/src/Resources/config/parameters_common.yaml [hunk 19]
```diff
@@ -8,8 +8,7 @@ parameters:
     # Administration locale. Must be registered in domains.yaml
     shopsys.admin_locale: en
 
-    # All time values are displayed in this timezone (default is PHP timezone)
-    shopsys.display_timezone: ~
+    shopsys.admin_display_timezone: 'Europe/Prague'
```
Comment (grossmannmartin) [inline packages/framework/src/Resources/config/parameters_common.yaml]: I would like to have a comment with explanation of this setting 😇

File: project-base/app/config/parameters_common.yaml [hunk 20]
```diff
@@ -21,7 +21,6 @@ parameters:
     container.autowiring.strict_mode: true
     container.dumper.inline_class_loader: true
     build-version: '0000000000000000_%kernel.environment%'
-    shopsys.display_timezone: Europe/Prague
     shopsys.cron_timezone: Europe/Prague
```
Comment (grossmannmartin) [inline project-base/app/config/parameters_common.yaml:24]: It seems kinda weird that we have one timezone set in the framework and the second (cron) is only defined in the framework and set in project-base.

File: packages/frontend-api/src/Model/Resolver/OpeningHours/OpeningHoursResolverMap.php [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+     * @param \Shopsys\FrameworkBundle\Model\Store\OpeningHours\OpeningHoursDataFactory $openingHoursDataFactory
+     * @param \Shopsys\FrameworkBundle\Component\Localization\DisplayTimeZoneProvider $displayTimeZoneProvider
+     * @param \Shopsys\FrameworkBundle\Component\DateTimeHelper\DateTimeHelper $dateTimeHelper
+     * @param \Shopsys\FrameworkBundle\Model\Store\OpeningHours\StoreOpeningHoursProvider $storeOpeningHoursProvider
+     */
+    public function __construct(
+        protected readonly Domain $domain,
+        protected readonly ClosedDayFacade $closedDayFacade,
+        protected readonly OpeningHoursDataFactory $openingHoursDataFactory,
+        protected readonly DisplayTimeZoneProviderInterface $displayTimeZoneProvider,
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Resolver/OpeningHours/OpeningHoursResolverMap.php]: those four dependencies seems unused now

File: packages/framework/src/Resources/translations/validators.cs.po [hunk 22]
```diff
@@ -127,6 +127,9 @@ msgstr "Počet dní k expedici musí být větší nebo roven {{ compared_value
 msgid "Number of days to expedite must be whole number."
 msgstr "Počet dní k expedici musí být celé číslo"
 
+msgid "Opening hours setting is not valid"
+msgstr "Nastavení otevírací doby není validní"
```
Comment (grossmannmartin) [inline packages/framework/src/Resources/translations/validators.cs.po]:
```suggestion
msgstr "Neplaté nastavení otevírací doby"
```

maybe?

File: packages/framework/src/Component/DateTimeHelper/DateTimeHelper.php [hunk 23]
```diff
@@ -78,4 +51,25 @@ public static function getCurrentRoundedTimeForIntervalAndTimezone(
...
+    public static function createDateTimeFromTime(string $hoursAndMinutes): DateTimeImmutable
+    {
+        return new DateTimeImmutable(sprintf('1970-01-01 %s:00', $hoursAndMinutes));
+    }
+
+    /**
+     * @param int $domainId
+     * @return int
+     */
+    public function getDayOfWeek(int $domainId): int
```
Comment (grossmannmartin) [inline packages/framework/src/Component/DateTimeHelper/DateTimeHelper.php]: maybe `getCurrentDayOfWeek`?

File: project-base/app/src/Form/Admin/PromoCodeFormTypeExtension.php [hunk 24]
```diff
@@ -181,28 +180,20 @@ private function buildTimeValidationFormGroup(FormBuilderInterface $builder): vo
             'label' => t('Apply according to date and time limit'),
         ]);
 
-        $timeValidationGroup->add('dateValidFrom', DatePickerType::class, [
-            'view_timezone' => DateTimeHelper::UTC_TIMEZONE,
+        $timeValidationGroup->add('datetimeValidFrom', DateTimeType::class, [
```
Comment (grossmannmartin) [inline project-base/app/src/Form/Admin/PromoCodeFormTypeExtension.php:183]: Can you please notify the tester to test promo codes?

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/framework/src/Form/Admin/Store/StoreFormType.php (hunk 4 above)
Comment (vitek-rostislav) [reply packages/framework/src/Form/Admin/Store/StoreFormType.php]: that is a good catch! Even more, the JS was already broken by another missing callback validation, fixed in 3be2397433

File: packages/framework/src/Form/DatePickerType.php (hunk 5 above)
Comment (vitek-rostislav) [reply packages/framework/src/Form/DatePickerType.php]: true, fixed in 958cbcbe3e, 87d99c5dd4, and c759934ca9

File: packages/framework/src/Form/DatePickerType.php (hunk 6 above)
Comment (vitek-rostislav) [reply packages/framework/src/Form/DatePickerType.php]: good catch, fixed in 1722d2beed

File: packages/framework/src/Migrations/Version20240131163656.php (hunk 8 above)
Comment (vitek-rostislav) [reply packages/framework/src/Migrations/Version20240131163656.php:38]: :ok_hand: 764c4cda64

File: packages/framework/src/Migrations/Version20240130104139.php (hunk 7 above)
Comment (vitek-rostislav) [reply packages/framework/src/Migrations/Version20240130104139.php:18]: :ok_hand: 764c4cda64

File: packages/framework/src/Migrations/Version20240131163656.php (hunk 8 above)
Comment (vitek-rostislav) [reply packages/framework/src/Migrations/Version20240131163656.php]: yeah, there might be entries entry with NULL opening hours for a day, see the screens below:
image
image

So after the migration, you end up with a row containing null values, however, now I am not sure why this was a problem :smile: I will test the migration one more time.

File: packages/framework/src/Form/Transformers/OpeningHoursCollectionTransformer.php (hunk 9 above)
Comment (vitek-rostislav) [reply packages/framework/src/Form/Transformers/OpeningHoursCollectionTransformer.php]: well, unlike the other method, this code is used only here, but I am ok with moving it there (I actually like the result as it puts a name on the functionality that was otherwise anonymous in the `reverseTransform` method) 88fdeb53ef

File: packages/framework/src/Model/Store/StoreDataFactory.php (hunk 10 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/StoreDataFactory.php]: fair point, fixed in 3aaee55ef9

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php (hunk 11 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: nice! 07c7a4f988

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php (hunk 12 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php:28]: I was thinking about that during the implementation and thought it was unnecessary here. On the other hand, now, with some time gap, I think it won't harm and maybe it will save someone from unexpected behavior in the future, so added in 5cee258c00

File: packages/framework/src/Form/DatePickerType.php (hunk 6 above)
Comment (vitek-rostislav) [reply packages/framework/src/Form/DatePickerType.php]: aaand in 5d254979e1 :smile:

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php (hunk 13 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: great idea, reworked in 69985a44f9

File: packages/framework/src/Migrations/Version20240131163656.php (hunk 8 above)
Comment (vitek-rostislav) [reply packages/framework/src/Migrations/Version20240131163656.php]: solved by 69985a44f9

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php (hunk 14 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: the code has changed with the refactoring, but 9cbf232a34 :+1:

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php (hunk 15 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: makes sense - 093dad2fd8

File: packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php (hunk 16 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/OpeningHours/StoreOpeningHoursProvider.php]: great idea - 658d9162de

File: packages/framework/src/Model/Store/StoreFacade.php (hunk 17 above)
Comment (vitek-rostislav) [reply packages/framework/src/Model/Store/StoreFacade.php]: I like the `createWithStore` approach - b88577fa93

File: project-base/app/tests/FrontendApiBundle/Functional/Order/TransportInOrderValidationTest.php (hunk 18 above)
Comment (vitek-rostislav) [reply project-base/app/tests/FrontendApiBundle/Functional/Order/TransportInOrderValidationTest.php:111]: To me, it was an edge case - the admin would have to change the store domain in production when a customer had the store selected as a pickup store in his cart.

I could create a test for this behavior but honestly, I do not want to postpone the task with that anymore

File: packages/framework/src/Resources/config/parameters_common.yaml (hunk 19 above)
Comment (vitek-rostislav) [reply packages/framework/src/Resources/config/parameters_common.yaml]: added in 6330541347

File: project-base/app/config/parameters_common.yaml (hunk 20 above)
Comment (vitek-rostislav) [reply project-base/app/config/parameters_common.yaml:24]: lets discuss it please, if it was up to me, I would move the cron timezone value setting to the framework. But maybe it was [implemented](https://github.com/shopsys/shopsys/commit/262c9861) this way for a reason that I do not see at the moment.

File: packages/frontend-api/src/Model/Resolver/OpeningHours/OpeningHoursResolverMap.php (hunk 21 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Resolver/OpeningHours/OpeningHoursResolverMap.php]: :+1: I have removed them in 69985a44f9df3335fe416ec5f615bb790c69f383

File: packages/framework/src/Resources/translations/validators.cs.po (hunk 22 above)
Comment (vitek-rostislav) [reply packages/framework/src/Resources/translations/validators.cs.po]: thanks, that is definitely less developer-slangish :smile:

File: packages/framework/src/Component/DateTimeHelper/DateTimeHelper.php (hunk 23 above)
Comment (vitek-rostislav) [reply packages/framework/src/Component/DateTimeHelper/DateTimeHelper.php]: yes! :ok_hand: c360e0befa

File: project-base/app/src/Form/Admin/PromoCodeFormTypeExtension.php (hunk 24 above)
Comment (vitek-rostislav) [reply project-base/app/src/Form/Admin/PromoCodeFormTypeExtension.php:183]: sure thing :+1:

File: packages/framework/src/Model/Store/OpeningHours/OpeningHoursRange.php [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    public function getOpeningTime(): string
+    {
+        return $this->openingTime;
+    }
+
+    /**
+     * @return string
+     */
+    public function getClosingTime(): string
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Store/OpeningHours/OpeningHoursRange.php]: after rebase the check for not using types in entities should warn you about setters, getters, and properties in entity and data object. Can you please verify then?

File: packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+declare(strict_types=1);
+
+namespace Shopsys\FrontendApiBundle\Model\Store\OpeningHours;
+
+use DateTimeImmutable;
+use Shopsys\FrameworkBundle\Model\Store\OpeningHours\StoreOpeningHoursProvider;
+use Shopsys\FrameworkBundle\Model\Store\Store;
+
+class StoreOpeningHoursApiProvider extends StoreOpeningHoursProvider
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php]: I don't think this provider should extend the one from framework. 
I believe the composition is a better design pattern in this case.

And maybe some methods may be moved to frontend api package... or maybe only one storeopeninghoursprovider in framework is sufficient 🤔  

let's discuss it please

File: packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    /**
+     * @param \DateTimeImmutable $date
+     * @param \Shopsys\FrameworkBundle\Model\Store\Store $store
+     * @return \Shopsys\FrontendApiBundle\Model\Store\OpeningHours\OpeningHoursWithDateData
+     */
+    protected function getOpeningHoursDataForDate(DateTimeImmutable $date, Store $store): OpeningHoursWithDateData
+    {
+        $openingHoursForDay = $this->getOpeningHoursSetting($store)->forDate($date);
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php]: nitpick

`...ForDay` or `...ForDate`?

File: packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Store\Store $store
+     * @return \Shopsys\FrontendApiBundle\Model\Store\OpeningHours\OpeningHoursWithDateData[]
+     */
+    public function getFollowingWeekOpeningHours(Store $store): array
+    {
+        $openingHoursData = [];
+
+        $today = new DateTimeImmutable('today', timezone: $this->displayTimeZoneProvider->getDisplayTimeZoneByDomainId($store->getDomainId()));
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php]:
```suggestion
        $today = new DateTimeImmutable('today', $this->displayTimeZoneProvider->getDisplayTimeZoneByDomainId($store->getDomainId()));
```

timezone is a second argument, so named argument is unnecessary

File: packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+
+        for ($i = 0; $i <= 6; $i++) {
+            $day = $today->modify("+${i} days");
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php]: in PHP 8.3 the proper way is this
```suggestion
            $day = $today->modify("+{$i} days");
```

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php (hunk 26 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php]: b61c2061f9

File: packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php (hunk 27 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php]: yes! 3da408586a

File: packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php (hunk 28 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php]: thanks, resolved in https://github.com/shopsys/shopsys/commit/b61c2061f9673808794d71538399a24e737ca785

File: packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php (hunk 29 above)
Comment (vitek-rostislav) [reply packages/frontend-api/src/Model/Store/OpeningHours/StoreOpeningHoursApiProvider.php]: thank you! resolved in https://github.com/shopsys/shopsys/commit/b61c2061f9673808794d71538399a24e737ca785

File: project-base/app/config/parameters_common.yaml (hunk 20 above)
Comment (vitek-rostislav) [reply project-base/app/config/parameters_common.yaml:24]: unified in 468a5a17c1 :+1:

**Review by grossmannmartin — APPROVED** [review]

**Review by TomasLudvik — APPROVED** [review]

**Review by pk16011990 — APPROVED** [review]
